### DISCUSSION TITLE
TypeScript

### DISCUSSION THREAD

**Discussion by hydroper** [airsdk/Adobe-Runtime-Support#2595] [General]
Draft
https://github.com/hydroper/ts4air
Now the project's README is cleaner!

**Comment by teotigraphix** [comment]
Curious what your end goal is here. What are your own uses cases for this, more tooling/language features with TypeScript or some other reason?

Does this have to do with your VioletScript?

**Reply by hydroper** [reply]
Just more language features really. TypeScript just works and TypeDoc supports, say, Markdown. I like my language, though it's:

- A bit incompatible with ActionScript. It could still target AVM2 somehow.
- I'm not sure how reflection will be done when targetting AVM2 (e.g. type meta-objects).
- Its compiler is still in production.

TypeScript is also able to reuse NPM. And why create apps in Adobe AIR with TypeScript when you can do the same with a browser? Because of the file system advantages: `file:` scheme, `app:` scheme and `app-storage:` scheme. The storage available for browser apps is said to be very limited when using indexedDB, localStorage or cookies. I'll be honest I didn't use Adobe AIR a lot though.

**Comment by hydroper** [comment]
I'm beginning to work on this tool. To support `async` functions and their `await` operator I'll need `Promise`. However this could be part of the AIR API:

- I've a basic writer for ABC and SWF (very little, based on brion's `wasm2swf` tool) in the TypeScript tool, `ts4air`, but I don't have a reader for ABC and SWF yet.
- A typings module, `com.adobe.air` will declare globals, including `Promise`, but won't be able to provide implementation by itself. (I'm talking about the `declare global {...}` statement in a global `d.ts` file that every `ts4air` project will use.

I already have an implementation for `Promise` in AS3 that I made recently: https://github.com/rialight/api-adobeair/blob/master/com.rialight.util/src/com/rialight/util/Promise.as I can only use it if I load ABC from a SWF. As I'm editing my comment here, I've decided I'll try to do this.

If you've not noted it, I've updated the [project README](https://github.com/hydroper/ts4air).
